lib/evaluate/traverse.h

```diff
@@ -0,0 +1,311 @@
+    if (const auto *symbol{x.GetAssumedTypeDummy()}) {
+      return visitor_(*symbol);
+    } else {
+      return visitor_(x.UnwrapExpr());
+    }
+  }
+  Result operator()(const ProcedureRef &x) const {
+    return Combine(x.proc(), x.arguments());
+  }
+  template<typename T> Result operator()(const FunctionRef<T> &x) const {
+    return visitor_(static_cast<const ProcedureRef &>(x));
+  }
+
+  // Other primaries
+  template<typename T>
+  Result operator()(const ArrayConstructorValue<T> &x) const {
+    return visitor_(x.u);
+  }
+  template<typename T>
+  Result operator()(const ArrayConstructorValues<T> &x) const {
+    return CombineContents(x);
+  }
+  template<typename T> Result operator()(const ImpliedDo<T> &x) const {
+    return Combine(x.lower(), x.upper(), x.stride(), x.values());
+  }
+  Result operator()(const semantics::ParamValue &x) const {
+    return visitor_(x.GetExplicit());
+  }
+  Result operator()(
+      const semantics::DerivedTypeSpec::ParameterMapType::value_type &x) const {
+    return visitor_(x.second);
+  }
+  Result operator()(const semantics::DerivedTypeSpec &x) const {
+    return CombineContents(x.parameters());
+  }
+  Result operator()(const StructureConstructorValues::value_type &x) const {
+    return visitor_(x.second);
+  }
+  Result operator()(const StructureConstructor &x) const {
+    return visitor_.Combine(visitor_(x.derivedTypeSpec()), CombineContents(x));
+  }
+
+  // Operations and wrappers
+  template<typename D, typename R, typename O>
+  Result operator()(const Operation<D, R, O> &op) const {
+    return visitor_(op.left());
+  }
+  template<typename D, typename R, typename LO, typename RO>
+  Result operator()(const Operation<D, R, LO, RO> &op) const {
+    return Combine(op.left(), op.right());
+  }
+  Result operator()(const Relational<SomeType> &x) const {
+    return visitor_(x.u);
+  }
+  template<typename T> Result operator()(const Expr<T> &x) const {
+    return visitor_(x.u);
+  }
+
+private:
+  template<typename ITER> Result CombineRange(ITER iter, ITER end) const {
+    if (iter == end) {
+      return visitor_.Default();
+    } else {
+      Result result{visitor_(*iter++)};
+      for (; iter != end; ++iter) {
+        result = visitor_.Combine(std::move(result), visitor_(*iter));
+      }
+      return result;
+    }
+  }
+
+  template<typename A> Result CombineContents(const A &x) const {
+    return CombineRange(x.begin(), x.end());
+  }
+
+  template<typename A, typename... Bs>
+  Result Combine(const A &x, const Bs &... ys) const {
+    if constexpr (sizeof...(Bs) == 0) {
+      return visitor_(x);
+    } else {
+      return visitor_.Combine(visitor_(x), Combine(ys...));
+    }
+  }
+
+  Visitor &visitor_;
+};
+
+template<typename Visitor, typename Result> class TraverseBase {
+public:
+  explicit TraverseBase(Visitor &v) : traverse_{v} {}
+  static Result Default() { return {}; }
+  template<typename A> Result operator()(const A &x) const {
```

Comment:
This default template overload is the issue with `Operation<>` overloads. It has priority over down casts. I cannot find a clever SFINAE solution to prevent that.
The old framework, though more complex, did not have this issue.

Reply:
One in-between solution I can think of is to have another CRTP helper class to catch operations.
Something like:
```
template<typename Visitor>
class OperationHandler {
  using Result = typename Visitor::Result;
  // only here all ops must be defined
  template<typename T>
  Result operation()(const Add<T>& add) {
    return Visitor::HandleOperation(add);
  }
  // explicitly define operator for all other ops ....
};

class GetShapeHelper
  : public AnyTraverse<GetShapeHelper, std::optional<Shape>>, OperationHandler<GetShapeHelper> {
public: 
  // ....
  template<typename D, typename R, typename LO, typename RO>
  Result HandleOperation(const Operation<D, R, LO, RO> &operation) const {
    // ...
  }
};
```
Not sure it is a nice design pattern though.

Reply:
The solution turned out to be simple: just don't have a catch-all template in the visitor.

Reply:
The default general template turns out to not be necessary.